lib/policyEvaluator/RequestContext.js

```diff
@@ -0,0 +1,417 @@
+// For bucket head and object head:
+// http://docs.aws.amazon.com/AmazonS3/latest/dev/
+// using-with-s3-actions.html
+const _actionMap = {
+    bucketDelete: 's3:DeleteBucket',
+    bucketGet: 's3:ListBucket',
+    bucketGetACL: 's3:GetBucketAcl',
+    bucketHead: 's3:ListBucket',
+    bucketPut: 's3:CreateBucket',
+    bucketPutACL: 's3:PutBucketAcl',
+    completeMultipartUpload: 's3:PutObject',
+    initiateMultipartUpload: 's3:PutObject',
+    listMultipartUploads: 's3:ListBucketMultipartUploads',
+    listParts: 's3:ListMultipartUploadParts',
+    multipartDelete: 's3:AbortMultipartUpload',
+    objectDelete: 's3:DeleteObject',
+    objectGet: 's3:GetObject',
+    objectGetACL: 's3:GetObjectAcl',
+    objectHead: 's3:GetObject',
+    objectPut: 's3:PutObject',
+    objectPutACL: 's3:PutObjectAcl',
+    objectPutPart: 's3:PutObject',
+    serviceGet: 's3:ListAllMyBuckets',
+};
+
+
+function _findAction(service, method) {
+    if (service === 's3') {
+        return _actionMap[method];
+    }
+    return undefined;
+}
+
+function _buildArn(service, bucket, object) {
+    if (service === 's3') {
+        if (bucket && object) {
+            return `arn:aws:s3:::${bucket}/${object}`;
+        } else if (bucket) {
+            return `arn:aws:s3:::${bucket}`;
+        }
+        return 'arn:aws:s3:::';
+    }
+    return undefined;
+}
+
+/**
+ * Class containing RequestContext for policy auth check
+ * @param {object} headers - request headers
+ * @param {query} query - request query
+ * @param {string} bucketName - bucket name from request if any
+ * @param {string} objectName - object name from request if any
+ * @param {string} requesterIp - ip of requester
+ * @param {boolean} isEncrypted - whether request was https
+ * @param {string} apiMethod - type of request
+ * @param {string} awsService - service receiving request
+ * @param {string} locationConstraint - location constraint
+ * for put bucket operation
+ * @param {object} requesterInfo - info about entity making request
+ * @param {string} signatureVersion - auth signature type used
+ * @param {string} authType - type of authentication used
+ * @param {number} signatureAge - age of signature in milliseconds
+ * @return {RequestContext} a RequestContext instance
+ */
+
+class RequestContext {
+    constructor(headers, query, bucketName, objectName,
+        requesterIp, isEncrypted, apiMethod,
+        awsService, locationConstraint, requesterInfo,
+        signatureVersion, authType, signatureAge) {
+        this._headers = headers;
+        this._query = query;
+        this._requesterIp = requesterIp;
+        this._sslEnabled = isEncrypted;
+        this._apiMethod = apiMethod;
+        this._awsService = awsService;
+        this._bucket = bucketName;
+        this._object = objectName;
+        this._locationConstraint = locationConstraint;
+        // Not implemented
+        this._multiFactorAuthPresent = undefined;
```

Comment:
a `null` would make more sense than `undefined` for properties that are not set. This has zero impact on logic but conventional code uses it that way, so it's easy for someone new looking at the code.